### PR TITLE
[skunkworks] stash: Add a WireCompatible trait to reduce boilerplate in Stash migrations

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR adds a new `unsafe trait WireCompatible<T>` that defines an implementing type as being wire compatible with `T`, and a macro `wire_compatible!` that implements the trait for you. The macro also generates `proptest` cases where we create arbitrary instances of type A, serialize it, then deserialize the bytes with type B and assert it succeeded, to test the property that the two types are actually wire compatible.

The PR also slightly modifies the existing v27 to v28 migration to use the new macro and trait.

### Motivation

Writing would-be-simple Stash migrations where the new Stash version is exactly the same, or wire compatible, with the previous version is not actually simple, because you need to write a decent amount of boilerplate code of `impl From<OLD_VERSION> for NEW_VERSION`. This PR prevents the need to write any of that code, and instead we rely on the property that protobuf types are generally forward compatible with one another.

### Tips for reviewer

* Hide whitespace

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a

### CODE REVIEW

File: src/stash/src/objects.rs [hunk 1]
```diff
@@ -176,6 +176,60 @@ impl From<String> for proto::StringWrapper {
...
+
+                #[mz_ore::test]
+                #[cfg_attr(miri, ignore)] // slow
+                fn [<proptest_wire_compat_ $b:snake $(_$b_sub:snake)* _to_ $a:snake $(_$a_sub:snake)* >](b: $b $(::$b_sub)* ) {
+                    use ::prost::Message;
+                    let bytes = b.encode_to_vec();
+                    let decoded = $a $(::$a_sub)*::decode(&bytes[..]);
+                    ::proptest::prelude::prop_assert!(decoded.is_ok());
+                }
+            }
```
Comment (jkosh44) [inline src/stash/src/objects.rs:227]: Is is worth adding a test that does something like `assert_eq!(a.encode_to_vec(), b.encode_to_vec())`

**Review by jkosh44 — APPROVED** [review]
Awesome!

File: src/stash/src/objects.rs (hunk 1 above)
Comment (ParkMyCar) [reply src/stash/src/objects.rs:233]: Good thought, I added the assert!

FWIW I'm pretty sure this is true for protobuf, but there might be an edge case where this assert would fail, meanwhile the types are wire compatible. It definitely makes the test more strict though which IMO is a good idea, and we can relax the constraint in the future if need be!
